### DISCUSSION TITLE
In need of help creating menu item for Thinstation in ipxe

### DISCUSSION THREAD

**Discussion by StrubeNL** [ipxe/ipxe#1131] [General]
Hi all,
I've been testing for weeks now and going around in circles, never finding a correct answer.
We've been using ipxe as a way to networkboot an exam environment with a safe browser. 
Booting occurs on different kinds of HP hardware and with safe boot "off" they can boot into the iPXE menu and can select and boot the image. (pxe loads ipxe64.efi on boot, don't know if this is relevant)

The menu item is as follows:
```
:facetmhv
kernel http://[ipxeserver-ip-address]/facetmhv/vmlinuz initrd=initrd rhgb boot=casper ip=dhcp toram bcld.afname.url=https://[website-address] snd_hda_intel.dmic_detect=0 url=http://[ipxeserver-ip-address]/facetmhv/bcld.iso
initrd http://[ipxeserver-ip-address]/facetmhv/initrd
boot
```

Now what'really doing my head in is to create a new menu item for a bootable Thinstation client. For the love of god i cannot get it to work. I've created a working "all modules" build which i can load on a VMWare vm testmachine, but i want to use on a HP laptop, the same way we boot the other image.

What i've got now is this:

```
:thinstation_efi
kernel http://[ipxeserver-ip-address]/thin/pxe/boot/vmlinuz initrd=initrd
append http://[ipxeserver-ip-address]/thin/pxe/boot/initrd
boot

```

But it gives an error message "Exec format error" which doesn't help me in any kind of way.
Searching around i only find an answer that i should be loading Syslinux for EFI boot.
When i browse the "pxe" directory i doe have a efi64 subfolder which contains syslinux.efi etc but how do i get that into the menu item?

Like i said, any help pointing me in the right direction would be helpful!

**Comment by NiKiZe** [comment]
What is the full error message you get? Every error iPXE produces also contains an https://ipxe.org/.... Url have you followed that url? What does that page say?

My guess is that you need a kernel with EFI_STUB enabled. But the full error message with url would be good to verify. You can also insert an imgstat line before your boot.

**Reply by StrubeNL** [reply]
The full error message is as follows: 
```
http://[ipxeserver-ip-address]/thin/pxe/boot/vmlinuz... ok
Could not select: Exec format error (http://ipxe.org/2e008081)
Could not boot: Exec format error (http://ipxe.org/2e008081)
```

I'v exported the boot-images directory listing from Thinstation on [pastebin](https://pastebin.com/TYvU8Rwg), maybe i'm using the wrong kernel/initrd ? This export contains all boot options from Thinstation, i use the vmlinuz and initrd from Line 87 (pxe).

**Reply by NiKiZe** [reply]
Did you read http://ipxe.org/2e008081 ?
Which kernel version is that, is it compiled with EFI_STUB enabled?

Some don't think it is needed, especially old versions don't have it. But iPXE will only run EFI binaries, so if you need that version, you might have to recompile the kernel yourself.

**Comment by StrubeNL** [comment]
I've read the url but skipped the config_efi_stub part, as it loads the other menu entry just fine (which loads an iso and boots that), but maybe the selected pxe vmlinuz kernel doesn't support that.. how do i check this or find out how? I

Loading an iso and booting it (ie a live-cd) would also be an option, as my install/image is only 80-90MB. But doing it the normal way would definitely have my preference.

**Comment by mcb30** [comment]
@StrubeNL could you provide a public download link to the kernel that you are trying to boot?

**Reply by StrubeNL** [reply]
https://www.dropbox.com/scl/fi/ntewvpyg4hz7sg9uwdgo0/vmlinuz?rlkey=oz0ofdlui65pgr1v0nov3bjwx&dl=0
https://www.dropbox.com/scl/fi/kxfcy4veo9fhod17r0tyl/initrd?rlkey=sva1wvmzetd16dsjxhq9z4dcm&dl=0

These are the vmlinuz and initrd mentioned in the menu item..

**Reply by mcb30** [reply]
Thanks.  The `vmlinuz` file appears to be corrupt.  Attempting to dump the PE headers fails:

```console
$ pe-dumpinfo vmlinuz\?rlkey\=oz0ofdlui65pgr1v0nov3bjwx\&dl\=0 
# file: vmlinuz?rlkey=oz0ofdlui65pgr1v0nov3bjwx&dl=0
Traceback (most recent call last):
  File "/usr/bin/pe-dumpinfo", line 33, in <module>
    sys.exit(load_entry_point('virt-firmware==23.11', 'console_scripts', 'pe-dumpinfo')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/virt/peutils/peutils.py", line 283, in pe_dumpinfo
    pe = pefile.PE(filename)
         ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pefile.py", line 2894, in __init__
    self.__parse__(name, data, fast_load)
  File "/usr/lib/python3.12/site-packages/pefile.py", line 3062, in __parse__
    raise PEFormatError("Invalid NT Headers signature.")
pefile.PEFormatError: 'Invalid NT Headers signature.'
```

Looking into the file contents, the NT headers are present at offset 0x80.  However, the `e_lfanew` field in the DOS headers (which points to the NT headers) has the value 0x82 instead of 0x80.

Using hexedit to change this value from 0x82 to 0x80 allows `pe-dumpinfo` to dump the PE headers.

I don't know how you've ended up with a kernel that has a one-byte corruption in it, but I would definitely throw away that kernel binary and get hold of a new one!
